File: tests/solutions_cpp/subtree_of_another_tree.cpp
#include "json_helper.h"

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!subRoot) return true;
        if (!root) return false;
        if (sameTree(root, subRoot)) return true;
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }
private:
    bool sameTree(TreeNode* a, TreeNode* b) {
        if (!a && !b) return true;
        if (!a || !b) return false;
        if (a->val != b->val) return false;
        return sameTree(a->left, b->left) && sameTree(a->right, b->right);
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto root = build_tree(args[0]);
    auto subRoot = build_tree(args[1]);
    bool result = sol.isSubtree(root, subRoot);
    cout << (result ? "true" : "false") << endl;
    return 0;
}

File: tests/solutions_js/subtree_of_another_tree.js
class Solution {
    isSubtree(root, subRoot) {
        if (subRoot === null) return true;
        if (root === null) return false;
        if (this._sameTree(root, subRoot)) return true;
        return this.isSubtree(root.left, subRoot) || this.isSubtree(root.right, subRoot);
    }

    _sameTree(p, q) {
        if (p === null && q === null) return true;
        if (p === null || q === null) return false;
        return p.val === q.val && this._sameTree(p.left, q.left) && this._sameTree(p.right, q.right);
    }
}
